doc/kernel-limitations.md

```diff
@@ -0,0 +1,174 @@
+separation between the processes, but the child or the parent would have
+no handle to reference the other side's memory.
+
+On the other hand, a child process created with `fork` is supposed to
+share variables with the parent. It's significantly more challenging to
+emulating a forked child process as a thread without underlying support
+of address space separation.
+
+
+## Limitations arisen from SGX's thread model
+
+With SGX1, the number of threads running inside an enclave, i.e.,
+**ethread**, has to be statically declared. Once the application starts,
+no more ethreads can be dynamically created. Meanwhile, the stack size
+of each ethread is also statically declared and allocated while creating
+the enclave. An application attempting to create more ethreads than what
+is declared will crash.
+
+A Mystikos application declares how many user threads are needed in
+config.json. A potential mitigation of the limitation is to make the setting big
+enough. The caveat is that we should ensure the stack size for each ethread
+is reasonable so we don't waste too much stack memory on unused threads.
+
+When an user application calls `pthread_create`, Mystikos makes an OCALL
+to the host to create a host thread, and re-enters the enclave with an
+ECALL, and once inside the enclave, jump to the thread routine. The long
+chain of events leads to delays and performance problems occasionally.
+
+The security implication of using host threads is that the untrusted host
+controls the thread scheduling, and may decide to starve or switch execution
+order of ethreads.
+
+The security and performance impact of this limitation can be mitigated with
+an enclave-resident thread scheduler, which uses M:N mapping to map ethreads
+to user application pthreads, and performs context switching between pthreads.
+
+This limitation is likely to be relaxed when EDMM of SGX2 is officially
+supported in Mystikos.
+
+## Limitations arisen from SGX's memory model
+
+With SGX1, the heap size of the enclave application has to be statically
+declared. Once the application starts, allocating more memory than what's
+declared causes a crash due to out of memory.
+
+A Mystikos application declares how big the heap size is needed in config.json.
+A potential mitigation of the limitation is to make the setting big enough.
+The downside of doing that is: 1) the SGX runtime will take more time to
+add/extend heap pages one by one during the initialization, and 2) The heap
+size is backed by physical EPC memory of the system. If the declared heap
+size is much more than the available physical EPC memory size, the application
+performance is likely to suffer due to severe `EPC paging`.
+
+Also with SGX1, the untrusted host can manipulate the page permissions at
+will because the page tables are solely controlled by the host. The implies
+the host can trigger page fault whenever and/or at whichever address it likes.
```

Comment:
```suggestion
```
In both SGX thread model and many other TEEs' thread model, page access related, #PF or EPT fault can be triggered anytime by the privileged system SW outside of the TCB.